### PR TITLE
Eric/movie ssr conversion

### PR DESCRIPTION
**Pull request by erc08**
- Added SSR for fetching movie data
- Fixed page reload issues
- Restructured layout files (header, nav bar, footer)

### CODE REVIEW

File: components/Layout/Layout.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+// Styles
+import styles from "./Layout.module.css";
+
+export default function Layout(props) {
+  return (
+    <div className={styles.container}>
+      <Header />
+      <Nav />
+      {props.children}
```
Comment (reco) [inline components/Layout/Layout.tsx]:
```suggestion
export default function Layout({children}) {
  return (
    <div className={styles.container}>
      <Header />
      <Nav />
      {children}
```

File: components/FavoriteButton/FavoriteButton.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// Libraries
+import React, { useState, useEffect, useMemo } from "react";
```
Comment (reco) [inline components/FavoriteButton/FavoriteButton.tsx]:
```suggestion
import React, { useState, useEffect } from "react";
```
if not needed remove

File: components/FavoriteButton/FavoriteButton.module.css [hunk 3]
```diff
@@ -1,18 +1,17 @@
 /* Standard button format */
 .button {
-  color: black;
-  background-color: white;
+  color: var(--text-color);
```
Comment (reco) [inline components/FavoriteButton/FavoriteButton.module.css]: great you are using css vars!

File: components/FavoriteButton/FavoriteButton.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  useEffect(() => {
+    setHasMounted(true);
+    // console.log("Favorites Button Mounted.");
+  }, []);
+
+  console.log(`- Movie [ ${id} ]: ${isFav} -`);
+
+  return (
+    hasMounted && (
```
Comment (reco) [inline components/FavoriteButton/FavoriteButton.tsx:35]: any reason you want to render the button only on the client? the Dom should be in sync client and server... server can just use the default state. client then hydrates what ever state is in the local storage

File: components/FavoriteButton/FavoriteButton.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      <button
+        onClick={clickHandler}
+        className={cx([
+          styles.button,
+          styles[`variant__${variant}`],
+          { [styles.isFav]: isFav },
+        ])}
```
Comment (reco) [inline components/FavoriteButton/FavoriteButton.tsx:42]: beautiful

File: components/FavoriteButton/useToggleFav.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import { useState } from "react";
```
Comment (reco) [inline components/FavoriteButton/useToggleFav.tsx:1]: this `useTobbleFav` does not match my code in `reco/#7-favorites-button`

File: components/Layout/Footer.tsx [hunk 7]
```diff
@@ -0,0 +1,20 @@
+// Libraries
+import React from "react";
+
+// Styles
+import styles from "./Layout.module.css";
+
+function Footer() {
+  console.log("Load Footer");
```
Comment (reco) [inline components/Layout/Footer.tsx]:
```suggestion
```

File: components/Layout/Footer.tsx [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+      <a>A website for browsing movies.</a>
+
+      <a href="https://github.com/ericc0429">
+        Created by Eric Chen (ericc0429).
+      </a>
+    </footer>
+  ); // End return
+}
+
+export default React.memo(Footer);
```
Comment (reco) [inline components/Layout/Footer.tsx]: You can do this but this component is so small and simple almost not worth it. You might add more over had then performance gains. Good write up: 

https://felixgerschau.com/react-performance-react-memo/

File: components/Layout/Header.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import Link from "next/link";
+
+// Components
+import Nav from "./Nav";
+
+// Styles
+import styles from "./Layout.module.css";
+
+function Header(/*  { pageName }: any  */) {
+  console.log("Load Header");
```
Comment (reco) [inline components/Layout/Header.tsx]:
```suggestion
```

File: components/Layout/Layout.tsx [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+export default function Layout({ children }) {
+  return (
+    <>
+      <Header />
+      <main className={styles.container}>{children}</main>
+      <Footer />
+    </>
+  );
+}
```
Comment (reco) [inline components/Layout/Layout.tsx]: Nice!

File: components/Layout/Nav.tsx [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import Link from "next/link";
+
+// Components
+import { Search } from "components/Search/";
+
+// Styles
+import styles from "./Layout.module.css";
+
+export default function Nav() {
+  console.log("Load Nav");
```
Comment (reco) [inline components/Layout/Nav.tsx]:
```suggestion
```
we never want to see a console.log in the code

File: components/Layout/Nav.tsx [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  return (
+    <div className={styles.navbar}>
+      <Link href="/movie">
+        <a className={styles.navcard}>Trending Movies</a>
+      </Link>
+
+      <Search />
+
+      {/* <Link href="/search">
```
Comment (reco) [inline components/Layout/Nav.tsx]: wanna bring this back?

File: components/Movie/Movie.tsx [hunk 13]
```diff
@@ -1,128 +1,40 @@
 // Libraries
 import React, { useState, useEffect } from "react";
+import { GetServerSidePropsContext } from "next";
 
 // Functions
-import { Favorites } from "components/Favorites";
+import Moviecard from "./Moviecard";
```
Comment (reco) [inline components/Movie/Movie.tsx]: nice

File: components/Trending/Trending.tsx [hunk 14]
```diff
@@ -1,58 +1,52 @@
...
-
+import styles from "./Trending.module.css";
 
 // API data
-const api_key="6c0d3427fcf17ddf80d04106c35a3a98";
-const url = `https://api.themoviedb.org/3/trending/movie/day?api_key=${ api_key }`;
-
-export default function Trending() {
+const api_key = "6c0d3427fcf17ddf80d04106c35a3a98";
+const url = `https://api.themoviedb.org/3/trending/movie/day?api_key=${api_key}`;
```
Comment (reco) [inline components/Trending/Trending.tsx]: wanna put this into env variables

```
API_KEY: 
API_URL: 
```

https://nextjs.org/docs/basic-features/environment-variables

File: components/Trending/Trending.tsx [hunk 15]
```diff
@@ -1,58 +1,52 @@
...
+    try {
+      // Parse data and store into movies array
+      const res = await fetch(url);
+      const data = await res.json();
+      // console.log(data.results);
+      setMovies(data.results);
+    } catch (err) {
+      console.log(err);
+    }
+  };
```
Comment (reco) [inline components/Trending/Trending.tsx]: wanna do this with `getServerSideProps `?

File: components/Movie/Moviecard.tsx [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+import { FavoriteButton } from "components/FavoriteButton";
+
+// Styles
+import styles from "components/Movie/Movie.module.css";
+
+interface IMoviecardProps {
+  moviedata: string;
+}
+
+export default function Moviecard({ moviedata }: IMoviecardProps) {
```
Comment (reco) [inline components/Movie/Moviecard.tsx]: Data should not come in here as a JSON string. Make sure you make a nice interface for movie.
Also Movie just renders `Movicard. What do you need `Movie` for?

File: components/Movie/Moviecard.tsx [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+  "November ",
+  "December ",
+];
+
+function parseReleaseDate(rawDate: string) {
+  let dateArr = rawDate.split("-");
+  let month = months[parseInt(dateArr[1])];
+  let date = month + parseInt(dateArr[2]).toString() + ", " + dateArr[0];
+  return date;
+}
```
Comment (reco) [inline components/Movie/Moviecard.tsx]: you could move this to `/utils` and then import it

File: styles/globals.css [hunk 18]
```diff
@@ -1,16 +1,37 @@
...
+  --border-color: #8899a6;
+  --card-color: #282828;
+  --text-color: #fafafa;
+  --alt-text-color: #aaaaaa;
+  --highlight-color: #2571ff;
+  --alt-highlight-color: #888888;
+  --fav-color: #ff5f5f;
+  --accent-color: #ffffff;
+}
+
```
Comment (reco) [inline styles/globals.css]: cool

File: components/Search/Search.tsx [hunk 19]
```diff
@@ -1,81 +1,38 @@
 // Libraries
-import React, { useState } from 'react';
+import React, { useState } from "react";
 
 // Functions
-import { Results } from 'components/Results';
-import { Trending } from 'components/Trending';
+import { Results } from "components/Results";
+import { Trending } from "components/Trending";
```
Comment (reco) [inline components/Search/Search.tsx]: not used

File: components/Search/Search.tsx [hunk 20]
```diff
@@ -1,81 +1,38 @@
...
-
-                <Trending />
-
-            </main>
-        </>
-    );
-}
\ No newline at end of file
+import sbstyles from "./Searchbar.module.css";
+import styles from "./Search.module.css";
```
Comment (reco) [inline components/Search/Search.tsx]: not used

File: components/Search/Searchbar.tsx [hunk 21]
```diff
@@ -0,0 +1,81 @@
+// Libraries
+import React, { useState } from "react";
+import { useRouter } from "next/router";
+
+// Functions
+import { Results } from "components/Results";
```
Comment (reco) [inline components/Search/Searchbar.tsx]: not used

File: components/Search/Searchbar.tsx [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+// Styles
+import styles from "./Searchbar.module.css";
+
+function Searchbar() {
+  const [query, setQuery] = useState("");
+
+  // const [movies, setMovies] = useState([]);
```
Comment (reco) [inline components/Search/Searchbar.tsx]: console.log

File: components/Search/Searchbar.tsx [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+        const data = await res.json(); // Parses into JSON format
+        setMovies(data.results); // Stores the results of the API fetch response into movies array
+      } else {
+        console.log("-- No Query Provided --"); // Debugging purposes
+        setMovies(undefined);
+      }
+    } catch (err) {
+      console.error(err);
+    }
+  }; // End searchMovies */
```
Comment (reco) [inline components/Search/Searchbar.tsx]: ?

File: components/Search/index.tsx [hunk 24]
```diff
@@ -1 +1 @@
-export { default as Search } from './Search';
```
Comment (reco) [inline components/Search/index.tsx:1]: nice linter working?

File: pages/search/[query].tsx [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+import { useRouter } from "next/router";
+import { GetServerSidePropsContext } from "next";
+
+// Functions
+import { Movie } from "components/Movie";
+import { Results } from "components/Results";
+
+// API data
+const api_key = "6c0d3427fcf17ddf80d04106c35a3a98";
+const api_url = `https://api.themoviedb.org/3/search/movie?api_key=${api_key}`;
```
Comment (reco) [inline pages/search/[query].tsx]: env variables

File: pages/search/[query].tsx [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+  //const router = useRouter();
+  //const { id } = router.query;
+  //console.log(id);
+  //console.log(JSON.stringify(data));
+  //return <Movie moviedata={JSON.stringify(data)} />; // End return
+}
+
+export async function getServerSideProps(context: GetServerSidePropsContext) {
+  const { query } = context.params;
+  console.log(query);
```
Comment (reco) [inline pages/search/[query].tsx]: console.log

File: pages/search/[query].tsx [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+
+  const url = api_url.concat(`&query=${query}`);
+
+  console.log("-- Submitting Search Query --"); // Debugging purposes
```
Comment (reco) [inline pages/search/[query].tsx]: console.log

**Comment by reco** [comment]
hey @ericc0429 [public/searchIcon512px.png](https://github.com/ericc0429/moviebrowser/pull/19#diff-96dd2a3db0ab764446c621a45841519f4b63ee5a2925b996a589d747404f9757) an sag

File: components/Layout/Nav.tsx (hunk 12 above)
Comment (erc08) [reply components/Layout/Nav.tsx]: nope!

File: components/FavoriteButton/useToggleFav.tsx (hunk 6 above)
Comment (erc08) [reply components/FavoriteButton/useToggleFav.tsx:1]: I initially had your code, but I tweaked it slightly in terms of parameters and where some checks were.

File: components/MovieCard/IMovieData.tsx [hunk 28]
```diff
@@ -0,0 +1,57 @@
+interface IProdCompany {
```
Comment (reco) [inline components/MovieCard/IMovieData.tsx]: - For interface declaration you do not need JSX. So change the extension from `tsx` to `ts`
- These are types not data. Call the file `MovieCard.types.ts`

File: components/MovieCard/index.tsx [hunk 29]
```diff
@@ -0,0 +1,3 @@
+export { default as MovieCard } from "./MovieCard";
+export * from "./IMovieData";
+export * from "./MovieCardList";
```
Comment (reco) [inline components/MovieCard/index.tsx]: don't you want to move MovieList to a new dir in components?

File: pages/movie/[id].tsx [hunk 30]
```diff
@@ -2,14 +2,16 @@
...
+import { MovieCard } from "components/MovieCard";
+import { IMovie } from "components/MovieCard";
 
 // API data
 const api_key = process.env.API_KEY;
 const api_url = process.env.API_URL_MOVIE;
 
 export default function Main({ data }) {
-  return <Movie moviedata={JSON.stringify(data)} />; // End return
+  console.log(data);
```
Comment (reco) [inline pages/movie/[id].tsx]: damn... 

```suggestion
```

File: components/MovieCard/MovieCard.tsx [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+// Styles
+import styles from "./MovieCard.module.css";
+
+interface IMovieProps {
+  movie: IMovie;
+  variant: "details" | "grid" | "list";
+}
+
+export default function MovieCard({ movie, variant }: IMovieProps) {
+  console.log("Moviecard");
```
Comment (reco) [inline components/MovieCard/MovieCard.tsx]:
```suggestion
```
damn

File: pages/search/[query].tsx [hunk 32]
```diff
@@ -2,13 +2,15 @@
...
-import { Results } from "components/Results";
+import Results from "components/Results";
+import { IMovie } from "components/MovieCard";
 
 // API data
 const api_key = process.env.API_KEY;
 const api_url = process.env.API_URL_SEARCH;
 
 export default function Main({ data }) {
+  console.log(typeof data);
```
Comment (reco) [inline pages/search/[query].tsx]:
```suggestion
```
damn

File: components/MovieCard/IMovieData.tsx [hunk 33]
```diff
@@ -17,6 +17,10 @@ interface IMovieGenre {
   name: string;
 }
 
+/* Movie Data Interface ==================
```
Comment (reco) [inline components/MovieCard/IMovieData.tsx]: - For interface declaration you do not need JSX. So change the extension from tsx to ts
- These are types not data. Call the file `MovieCard.types.ts`

**Review by reco — APPROVED** [review]

**Pull request merged**